src/redisraft.c

```diff
@@ -713,62 +713,154 @@ static int cmdRaftCluster(RedisModuleCtx *ctx, RedisModuleString **argv, int arg
  *   +OK
  *   -ERR error description
  */
-
 static int cmdRaftShardGroup(RedisModuleCtx *ctx, RedisModuleString **argv, int argc)
 {
-    RaftReq *req;
+    RedisRaftCtx *rr = &redis_raft;
 
     if (argc < 2) {
         RedisModule_WrongArity(ctx);
         return REDISMODULE_OK;
     }
 
-    if (!redis_raft.config->sharding) {
-        RedisModule_ReplyWithError(ctx, "ERR: RedisRaft sharding not enabled");
+    if (!rr->config->sharding) {
+        RedisModule_ReplyWithError(ctx, "ERR RedisRaft sharding not enabled");
+        return REDISMODULE_OK;
+    }
+
+    if (checkRaftState(rr, ctx) == RR_ERROR ||
+        checkLeader(rr, ctx, NULL) == RR_ERROR) {
         return REDISMODULE_OK;
     }
 
     size_t cmd_len;
     const char *cmd = RedisModule_StringPtrLen(argv[1], &cmd_len);
 
     if (!strncasecmp(cmd, "GET", cmd_len)) {
-        req = RaftReqInit(ctx, RR_SHARDGROUP_GET);
-        handleShardGroupGet(&redis_raft, req);
+        ShardGroup *sg = getShardGroupById(rr, rr->log->dbid);
```

Comment:
i feel including the whole former "handle" function inline here, makes this function way too big....  why not just call the handle function directly with just a slight change in prototype?

Reply:
The reason is there is no abstraction between raft.c and redisraft.c anymore. Having separate functions in raft.c doesn't make much sense. Maybe we should leave raft related functions in raft.c eventually. Also, these generic handle...() family functions pollutes namespace and against coding convention I believe (Public "connection" functions start with Conn, Shard family functions start with Shard... etc.). So, I'm looking to get rid of these handle functions as much as possible. 

Anyway, moved Shardgroup operations into cluster.c. It makes sense to keep sharding related functions together and makes the function shorter/more readable as you suggested I believe.